File: Maximum sum Rectangle - GFG/maximum-sum-rectangle.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {
            int N, M, x, y;
            String S[] = read.readLine().split(" ");
            N = Integer.parseInt(S[0]);
            M = Integer.parseInt(S[1]);
            int a[][] = new int[N][M];
            for (int i = 0; i < N; i++) {
                String s[] = read.readLine().split(" ");
                for (int j = 0; j < M; j++) a[i][j] = Integer.parseInt(s[j]);
            }
            Solution ob = new Solution();
            System.out.println(ob.maximumSumRectangle(N, M, a));
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    
    private int kadane(int sum[])
    {
        int n = sum.length;
        int max_so_far = -1000000000;
        int curr = 0;
        
        for(int i = 0; i < n; i++)
        {
            curr += sum[i];
            if(curr < 0)
            {
                max_so_far = Math.max(max_so_far, curr);
                curr = 0;
            }
            else
                max_so_far = Math.max(max_so_far, curr);
        }
        
        return max_so_far;
    }
    
    int maximumSumRectangle(int n, int m, int mat[][]) {
        
        int sum[] = new int[n];
        int ans = Integer.MIN_VALUE;
        
        for(int c1 = 0; c1 < m; c1++)
        {
            Arrays.fill(sum, 0);
            for(int c2 = c1; c2 < m; c2++)
            {
                for(int r = 0; r < n; r++)
                {
                    sum[r] += mat[r][c2];
                }
                
                int currSum = kadane(sum);
                ans = Math.max(ans, currSum);
            }
            
        }
        
        return ans;
    }
};

File: Maximum sum Rectangle - GFG/maximum-sum-rectangle.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution {
    
    int kadane(vector<int> &arr) {
        int mini = INT_MIN;
        for(int i = 0; i < arr.size(); i++) {
            mini = max(mini, arr[i]);
        }
        if(mini <= 0)
            return mini;
            
        int i = 0;
        int sum = 0;
        int ans = arr[0];
        while(i < arr.size()) {
            sum += arr[i];
            ans = max(ans, sum);
            if(sum < 0)
                sum = 0;
            i++;
        }
        
        return ans;
    }
    
  public:
    int maximumSumRectangle(int n, int m, vector<vector<int>> mat) {
        // code here
        int ans = INT_MIN;
        for(int c1 = 0; c1 < m; c1++) {
            
            vector<int> arr(n, 0);
            for(int c2 = c1; c2 < m; c2++) {
                for(int i = 0; i < n; i++) {
                    arr[i] += mat[i][c2];
                }
                ans = max(ans, kadane(arr));
            }
        }
        
        return ans;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N, M;
        cin >> N >> M;
        vector<vector<int>> v(N, vector<int>(M));
        for (int i = 0; i < N; i++)
            for (int j = 0; j < M; j++) cin >> v[i][j];
        Solution ob;
        cout << ob.maximumSumRectangle(N, M, v) << "\n";
    }
}
// } Driver Code Ends